### PR TITLE
Orthogonally decoupled variational strategy

### PR DESCRIPTION
**Pull request by gpleiss**
Implements [Orthogonally Decoupled Variational Gaussian Processes](https://arxiv.org/abs/1809.08820) from Salimbeni et al. (2018).

### CODE REVIEW

File: docs/source/variational.rst [hunk 1]
```diff
@@ -63,6 +63,13 @@ deterministic relationship between :math:`\mathbf f` and :math:`\mathbf u`.
    :members:
 
 
+:hidden:`OrthogonalDecoupledVariationalStrategy`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (jacobrgardner) [inline docs/source/variational.rst]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import torch
+
+from ..distributions import MultivariateNormal
+from ..utils.memoize import cached
+from ._variational_strategy import _VariationalStrategy
+
+
+class OrthogonalDecoupledVariationalStrategy(_VariationalStrategy):
+    r"""
+    As defined in `Salimbeni et al. (2018)`_.
```
Comment (jacobrgardner) [inline gpytorch/variational/orthogonal_decoupled_variational_strategy.py]:
```suggestion
    Implements orthogonally decoupled VGPs as defined in `Salimbeni et al. (2018)`_.
```

File: docs/source/variational.rst (hunk 1 above)
Comment (gpleiss) [reply docs/source/variational.rst]: I have another commit coming up (with some more docs) which does this.

**Comment by jacobrgardner** [comment]
Actual test failures are obviously unrelated to this PR -- probably something changed on pytorch master?

We should also fix our `add_` deprecation warnings before cutting our next minor release.

**Comment by gpleiss** [comment]
@jacobrgardner - yeah I'm fixing the test failures right now.

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+
+import torch
+
+from ..distributions import MultivariateNormal
+from ..utils.memoize import cached
+from ._variational_strategy import _VariationalStrategy
+
+
+class OrthogonalDecoupledVariationalStrategy(_VariationalStrategy):
```
Comment (jacobrgardner) [inline gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: nit: if we've come so far as to spell our `Orthogonal`, maybe we should pay the extra two characters to add `ly`

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    @property
+    @cached(name="prior_distribution_memo")
+    def prior_distribution(self):
+        out = self.model(self.inducing_points)
+        res = MultivariateNormal(out.mean, out.lazy_covariance_matrix.add_jitter())
+        return res
+
+    def forward(self, x, inducing_points, inducing_values, variational_inducing_covar=None):
+        if variational_inducing_covar is not None:
+            raise NotImplementedError("DecoupledVariationalStrategy currently works with DeltaVariationalDistribution")
```
Comment (jacobrgardner) [inline gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: This error message?

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        num_data = x.size(-2)
+        full_output = self.model(torch.cat([x, inducing_points], dim=-2))
+        full_mean = full_output.mean
+        full_covar = full_output.lazy_covariance_matrix
+
+        # Cache the kernel matrix with the cached CG calls
```
Comment (jacobrgardner) [inline gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: Is this comment right?

Comment (gpleiss) [reply gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: Good catch.

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py (hunk 4 above)
Comment (gpleiss) [reply gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: This is true (the variational strategy here is for the mean inducing points). But I can also put that error message in the constructor.

Comment (jacobrgardner) [reply gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: Sorry, my comment wasn't very clear. What I meant was:
```suggestion
            raise NotImplementedError("OrthogonalDecoupledVariationalStrategy currently works with DeltaVariationalDistribution")
```

**Comment by gpleiss** [comment]
@jacobrgardner #1058 fixes the error from pytorch#master

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py (hunk 3 above)
Comment (gpleiss) [reply gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: Done

File: gpytorch/variational/orthogonal_decoupled_variational_strategy.py (hunk 4 above)
Comment (gpleiss) [reply gpytorch/variational/orthogonal_decoupled_variational_strategy.py]: I kept the error message there, and added one in the constructor (and changed the class name in the error message to be correct).

**Comment by gpleiss** [comment]
@jacobrgardner all comments are addressed - and this latest commit should fix all the tests as well. Should be ready to go.

**Pull request merged**